File: windows/VoiceFlow.Core.Tests/TextJoinerTests.cs
using VoiceFlow.Core;

namespace VoiceFlow.Core.Tests;

public class TextJoinerTests
{
    [Fact]
    public void AddsSpaceBetweenSentenceAndWord()
    {
        Assert.Equal(" ", TextJoiner.Prefix("Adesso", new CursorContext.Character('!')));
        Assert.Equal(" ", TextJoiner.Prefix("Quando", new CursorContext.Character('a')));
        Assert.Equal(" ", TextJoiner.Prefix("Ciao", new CursorContext.Character('.')));
    }

    [Fact]
    public void NoSpaceAfterWhitespaceOrAtStart()
    {
        Assert.Equal("", TextJoiner.Prefix("Adesso", new CursorContext.Character(' ')));
        Assert.Equal("", TextJoiner.Prefix("Adesso", new CursorContext.Character('\n')));
        Assert.Equal("", TextJoiner.Prefix("Adesso", new CursorContext.AtStart()));
    }

    [Fact]
    public void NoSpaceWhenNextTextIsPunctuationOrEmpty()
    {
        Assert.Equal("", TextJoiner.Prefix(", però", new CursorContext.Character('o')));
        Assert.Equal("", TextJoiner.Prefix(".", new CursorContext.Character('o')));
        Assert.Equal("", TextJoiner.Prefix("", new CursorContext.Character('o')));
    }

    [Fact]
    public void NoSpaceAfterOpeningBracketOrQuote()
    {
        Assert.Equal("", TextJoiner.Prefix("ciao", new CursorContext.Character('(')));
        Assert.Equal("", TextJoiner.Prefix("ciao", new CursorContext.Character('"')));
        Assert.Equal("", TextJoiner.Prefix("ciao", new CursorContext.Character('«')));
    }

    [Fact]
    public void UnknownContextFallsBackToNothing()
    {
        Assert.Equal("", TextJoiner.Prefix("ciao", new CursorContext.Unavailable()));
    }

    [Fact]
    public void AccentedAndUppercaseStartsCountAsWords()
    {
        Assert.Equal(" ", TextJoiner.Prefix("È tardi", new CursorContext.Character('.')));
        Assert.Equal(" ", TextJoiner.Prefix("2024", new CursorContext.Character('l')));
    }
}

File: Tests/VoiceFlowCoreTests/TextJoinerTests.swift
import Testing
@testable import VoiceFlowCore

@Suite struct TextJoinerTests {
    @Test func addsSpaceBetweenSentenceAndWord() {
        #expect(TextJoiner.prefix(for: "Adesso", precededBy: .character("!")) == " ")
        #expect(TextJoiner.prefix(for: "Quando", precededBy: .character("a")) == " ")
        #expect(TextJoiner.prefix(for: "Ciao", precededBy: .character(".")) == " ")
    }

    @Test func noSpaceAfterWhitespaceOrAtStart() {
        #expect(TextJoiner.prefix(for: "Adesso", precededBy: .character(" ")) == "")
        #expect(TextJoiner.prefix(for: "Adesso", precededBy: .character("\n")) == "")
        #expect(TextJoiner.prefix(for: "Adesso", precededBy: .atStart) == "")
    }

    @Test func noSpaceWhenNextTextIsPunctuationOrEmpty() {
        #expect(TextJoiner.prefix(for: ", però", precededBy: .character("o")) == "")
        #expect(TextJoiner.prefix(for: ".", precededBy: .character("o")) == "")
        #expect(TextJoiner.prefix(for: "", precededBy: .character("o")) == "")
    }

    @Test func noSpaceAfterOpeningBracketOrQuote() {
        #expect(TextJoiner.prefix(for: "ciao", precededBy: .character("(")) == "")
        #expect(TextJoiner.prefix(for: "ciao", precededBy: .character("\"")) == "")
        #expect(TextJoiner.prefix(for: "ciao", precededBy: .character("«")) == "")
    }

    @Test func unknownContextFallsBackToNothing() {
        #expect(TextJoiner.prefix(for: "ciao", precededBy: .unavailable) == "")
    }

    @Test func accentedAndUppercaseStartsCountAsWords() {
        #expect(TextJoiner.prefix(for: "È tardi", precededBy: .character(".")) == " ")
        #expect(TextJoiner.prefix(for: "2024", precededBy: .character("l")) == " ")
    }
}
